File: SWEA/D3/1244. ［S／W 문제해결 응용］ 2일차 － 최대 상금/［S／W 문제해결 응용］ 2일차 － 최대 상금.cpp
#include <iostream>
#include <queue>
#include <cstring>
#include <unordered_map>
using namespace std;
int ans = 0;
int ab[10] = {0,};
int ten[6] = {100000, 10000, 1000, 100, 10, 1};
int swapnum(int num, int l, int r, int length) {
    int fix = 6 - length;
    int lnum = num % (ten[fix + l] * 10)/ ten[fix + l];
    int rnum = num % (ten[fix + r] * 10)/ ten[fix + r];
    num += lnum * (ten[fix + r] - ten[fix + l] );
    num += rnum * (ten[fix + l] - ten[fix + r] );

    return num;
}

bool issame(const string &a) {
    memset(ab,0,sizeof(ab));
    for (char ch : a) {
        ab[ ch - '0']++;
        if (ab[ch - '0'] % 2 == 0) return true;
    }
    return false;
}

int main(void) {
    ios_base::sync_with_stdio(false); cin.tie(0); cout.tie(0);
    int T; cin >> T;
    string a; int k;
    int len;
    unordered_map<int, int> v;
    for (int tc = 1; tc <= T; ++tc) {
        cin >> a >> k;
        len = (int) a.size();
        ans = 0;
        v.clear();
        queue<int> q;
        q.push(stoi(a));
        bool same_num = issame(a);
        v[q.front()] = 0;
        while (!q.empty())
        {
            int now_num = q.front();
            q.pop();
            if (v[now_num] == k) continue;
            for (int i = 0; i < len-1; ++i) {
                for (int k = i+1; k < len; ++k) {
                    int nxt_num = swapnum(now_num, i, k, len);
                    if (v.find(nxt_num) == v.end()) {
                        v[nxt_num] = v[now_num] + 1;
                        q.push(nxt_num);
                    }
                }
            }
        }
        for (auto &[i, e] : v) {
            if ((k-e)%2 == 0 || same_num) ans = max(ans, i);
        }
        cout << '#' << tc << ' ' <<ans <<'\n';
    }
    return 0;
}

File: SWEA/D3/1244. ［S／W 문제해결 응용］ 2일차 － 최대 상금/［S／W 문제해결 응용］ 2일차 － 최대 상금.py
from collections import deque
from collections import Counter

T = int(input())
for tc in range(1, T + 1):
    num, k = input().split()
    k = int(k)
    length = len(num)
    same_num = Counter(num).most_common(1)[0][1] >= 2  # 중복되는 숫자 존재하는지 판단
    v = {}  # empty dictionary
    dq = deque()  # BFS 탐색용
    dq.append((num, 0))  # 초기 탐색 입력.
    ans = -1
    while dq:
        now_num, now_swap = dq.popleft()  # 현재 탐색해야 하는 수, 이 수의 "최소 입력"
        if now_swap == k:  # swap 횟수가 K 인경우, 더이상 swap 불가능 
            continue
        lst = list(now_num)  # swap 용이하기 위해 리스트변환, str은 immutable한 문제 때문
        for i in range(length - 1):
            for j in range(i + 1, length):
                lst[i], lst[j] = lst[j], lst[i]  # swap
                a = ''.join(lst) 
                if a not in v:  # 방문 흔적 없으면 방문  및 swap 횟수 등록
                    v[a] = now_swap + 1
                    dq.append((a, now_swap + 1))  # 다음 swap 탐색하기 위해 queue 에 담음
                lst[i], lst[j] = lst[j], lst[i]  # 원복

    for key, val in v.items():
        # 같은 숫자가 존재하거나, 해당 수를 얻는데 있어,  k 에서 최소 swap 수를 뺀 결과가 짝수 인 경우 답 갱신
        if same_num:
            ans = max(ans, int(key))
        elif (k - val) % 2 == 0:
            ans = max(ans, int(key))
    print(f'#{tc} {ans}')